backend/app/balance_monitor.go

```diff
@@ -38,59 +73,317 @@ func (h *Handler) MonitorSystemBalanceAndHandleSettlement() {
 			continue
 		}
 
-		amountToTransfer := record.TFTAmount - record.TransferredTFTAmount
-		if systemTFTBalance < amountToTransfer {
+		if systemTFTBalance < record.TFTAmount {
 			logger.GetLogger().Warn().Msgf("Insufficient system balance to settle pending record ID %d", record.ID)
 			continue
 		}
 
-		if err = h.transferTFTsToUser(record.UserID, record.ID, amountToTransfer); err != nil {
-			logger.GetLogger().Error().Err(err).Send()
-			continue
+		if err = h.transferTFTsToUser(record.UserID, record.TFTAmount); err != nil {
+			logger.GetLogger().Error().Err(err).Msgf("Failed to settle pending record ID %d", record.ID)
+
+			transferState = models.FailedState
+			transferFailure = err.Error()
+		}
+
+		if err := h.db.UpdateTransferRecordState(record.ID, transferState, transferFailure); err != nil {
+			logger.GetLogger().Error().Err(err).Msgf("Failed to update pending record ID %d state", record.ID)
 		}
 	}
 
 	return nil
 }
 
-func (h *Handler) transferTFTsToUser(userID, recordID int, amountToTransfer uint64) error {
+func (h *Handler) transferTFTsToUser(userID int, amountToTransfer uint64) error {
 	user, err := h.db.GetUserByID(userID)
 	if err != nil {
-		return errors.Wrapf(err, "failed to get user for pending record ID %d", recordID)
+		return errors.Wrapf(err, "failed to get user %d", userID)
 	}
 
 	err = internal.TransferTFTs(h.substrateClient, amountToTransfer, user.Mnemonic, h.systemIdentity)
 	if err != nil {
-		return errors.Wrapf(err, "Failed to transfer TFTs for pending record ID %d", recordID)
+		return errors.Wrapf(err, "Failed to transfer TFTs to user %d", userID)
+	}
+
+	return nil
+}
+
+func (h *Handler) withdrawTFTsFromUser(userID int, userMnemonic string, amountToWithdraw uint64) error {
+	userIdentity, err := substrate.NewIdentityFromSr25519Phrase(userMnemonic)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to create identity for user %d", userID)
 	}
 
-	err = h.db.UpdatePendingRecordTransferredAmount(recordID, amountToTransfer)
+	err = internal.TransferTFTs(h.substrateClient, amountToWithdraw, h.config.SystemAccount.Mnemonic, userIdentity)
 	if err != nil {
-		return errors.Wrapf(err, "Failed to update transferred amount for pending record ID %d", recordID)
+		return errors.Wrapf(err, "Failed to transfer TFTs to user %d", userID)
 	}
 
 	return nil
 }
 
-func (h *Handler) notifyAdminWithPendingRecords(records []models.PendingRecord) error {
+func (h *Handler) fundUserToFulfillDiscount(ctx context.Context, user models.User, addedRentedNodes []types.Node, addedSharedNodes []kubedeployer.Node, discount discount) error {
+	// calculate resources usage in USD applying discount
+	// I took the cluster nodes since only the new node is in cluster.Nodes
+	dailyUsageInUSDMillicent, err := h.calculateResourcesUsageInUSDApplyingDiscount(ctx, user.ID, user.Mnemonic, addedRentedNodes, addedSharedNodes, discount)
```

Comment:
Please add a check that the user has a USD balance before starting calculating and funding him.